File: problem_0048.py
#!/usr/bin/env python3
"""
The series, 11 + 22 + 33 + ... + 1010 = 10405071317.
Find the last ten digits of the series, 11 + 22 + 33 + ... + 10001000.
"""

import time
import sys
import logging

logging.basicConfig(
    stream=sys.stderr,
    level=logging.INFO,
    format='%(levelname)s:%(message)s')

def main():
    num_sum = 0
    for num in reversed(range(1,1001)):
        digits = num**num % int(1e10)
        num_sum += digits
    num_sum = num_sum % int(1e10)
    print('Answer: {}'.format(num_sum))

if __name__ == "__main__":
    start = time.time()
    main()
    logging.info('That took {:4.2f} seconds'.format(time.time() - start))

File: problem_0048.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

import Test.HUnit ((@=?), assertBool, runTestTT, Test(..))
import Text.Printf (printf)
import System.Console.CmdArgs
import Data.Time (getCurrentTime, diffUTCTime)

-- The series, 1^1 + 2^2 + 3^3 + ... + 10^10 = 10405071317.
-- Find the last ten digits of the series, 1^1 + 2^2 + 3^3 + ... + 1000^1000.

data EulerArgs = 
    Euler 
    | AdHoc{ limit::Integer, digits::Int }
    | UnitTest
    deriving (Show, Data, Typeable)

problem0048 :: Integer -> Int -> Integer
problem0048 limit n = read $ last n $ show $ sum [n^n | n <- [1..limit]]
    where
        last n xs = reverse $ take n $ reverse xs

unitTests = map TestCase $
    []

exec :: EulerArgs -> IO ()
exec Euler = do
    let answer = problem0048 1000 10
    printf "Answer: %d\n" answer
exec AdHoc{..} = do
    print $ problem0048 limit digits
exec UnitTest = do 
    runTestTT $ TestList unitTests
    return ()

main :: IO ()
main = do
    args <- cmdArgs $ modes [
        Euler,
        AdHoc{ limit = 10, digits = 11 },
        UnitTest]
    start <- getCurrentTime
    exec args
    stop <- getCurrentTime
    printf "That took %s\n"  $ show $ diffUTCTime stop start
